src/game/logic/object/locomotor.cpp

```diff
@@ -676,60 +676,471 @@ void Locomotor::Loco_Update_Move_Towards_Position(
+            force3d.x = force * dir->x;
+            force3d.y = force * dir->y;
+            force3d.z = 0.0f;
+            physics->Apply_Motive_Force(&force3d);
+        }
+    }
 }
 
 void Locomotor::Move_Towards_Position_Wheels(
     Object *obj, PhysicsBehavior *physics, const Coord3D &goal_pos, float on_path_dist_to_goal, float desired_speed)
 {
-#ifdef GAME_DLL
-    Call_Method<void, Locomotor, Object *, PhysicsBehavior *, const Coord3D &, float, float>(
-        PICK_ADDRESS(0x004B9CD0, 0x0074FFC1), this, obj, physics, goal_pos, on_path_dist_to_goal, desired_speed);
-#endif
+    BodyDamageType damage = obj->Get_Body_Module()->Get_Damage_State();
+    float max_speed = Get_Max_Speed_For_Condition(damage);
+    float max_turn = Get_Max_Turn_Rate(damage);
+    float max_accel = Get_Max_Acceleration(damage);
+
+    if (desired_speed > max_speed) {
+        desired_speed = max_speed;
+    }
+
+    float turn_speed = m_template->m_minTurnSpeed;
+    float orientation = obj->Get_Orientation();
+    float goal_angle = GameMath::Atan2(goal_pos.y - obj->Get_Position()->y, goal_pos.x - obj->Get_Position()->x);
+
+    float turn_angle = Normalize_Angle(goal_angle, orientation);
+    bool backwards = false;
+
+    if (max_speed / 4.0f > turn_speed) {
+        turn_speed = max_speed / 4.0f;
+    }
+
+    float cur_speed = physics->Get_Forward_Speed_2D();
+    bool b = false;
+
+    if (cur_speed == 0.0f) {
+        Set_Flag(MOVING_BACKWARDS, false);
+
+        if (m_template->m_canMoveBackwards && GameMath::Fabs(turn_angle) > DEG_TO_RADF(90.0f)) {
+            Set_Flag(MOVING_BACKWARDS, true);
+            Set_Flag(FLAG_8, obj->Get_Geometry_Info().Get_Major_Radius() * 5.0f < on_path_dist_to_goal);
+        }
+    }
+
+    if (Get_Flag(MOVING_BACKWARDS)) {
+        if (GameMath::Fabs(turn_angle) < DEG_TO_RADF(90.0f)) {
+            backwards = false;
+            Set_Flag(MOVING_BACKWARDS, false);
+        } else {
+            backwards = true;
+            Set_Flag(FLAG_8, obj->Get_Geometry_Info().Get_Major_Radius() * 5.0f < on_path_dist_to_goal);
+            b = Get_Flag(FLAG_8);
+
+            if (!b) {
+                goal_angle = Normalize_Angle(goal_angle, DEG_TO_RADF(180.0f));
+                turn_angle = Normalize_Angle(goal_angle, orientation);
+            }
+        }
+    }
+
+    if (GameMath::Fabs(turn_angle) > DEG_TO_RADF(9.0f) && desired_speed > turn_speed) {
+        desired_speed = turn_speed;
+    }
+
+    if (backwards) {
+        cur_speed = -cur_speed;
+    }
+
+    float speed1 = cur_speed / Get_Braking() + 1.0f;
+    float speed2 = cur_speed / 1.5f * speed1 + cur_speed;
+    float speed3 = speed2;
+
+    if (speed2 < 10.0f) {
+        speed3 = 10.0f;
+    }
+
+    float speed4;
+    float orientation2;
+    float speed5;
+    float speed6;
+    float c;
+    float s;
+    Coord3D pos;
+    Coord3D pos2;
+
+    if (GameMath::Fabs(turn_angle) <= DEG_TO_RADF(15.0f)) {
+        goto l1;
```

Comment:
You can get rid of the goto by doing the following

```c++
bool is_small_turn = GameMath::Fabs(turn_angle) <= DEG_TO_RADF(15.0f)
```
And then put the next block of code until line 945 into condition:
```c++
if (!is_small_turn)
{
  speed4 = (desired_speed + cur_speed) * 15.0f / 2.0f;
  ...
  pos2.z = obj->Get_Position()->z;
}
```

And then add additional check to big condition below:

```c++
if (!is_small_turn && (!g_theAI->Get_Pathfinder()->Valid_Movement_Terrain(obj->Get_Layer(), this, &pos2)
        || !g_theAI->Get_Pathfinder()->Valid_Movement_Terrain(obj->Get_Layer(), this, &pos))) {
        ...
    } else {
     ...
    }

Reply:
Going to leave the goto as-is for now.